pkg/k8s/watchers/endpointsynchronizer.go

```diff
@@ -323,8 +354,31 @@ func deleteCEP(ctx context.Context, scopedLog *logrus.Entry, ciliumClient v2.Cil
 		scopedLog.Debug("Skipping CiliumEndpoint deletion because it has no k8s namespace")
 		return nil
 	}
-	if err := ciliumClient.CiliumEndpoints(namespace).Delete(ctx, podName, meta_v1.DeleteOptions{}); err != nil {
-		if !k8serrors.IsNotFound(err) {
+
+	// A CEP should be only be deleted by the agent that manages the
+	// corresponding pod. However, it is possible for a pod to restart and be
+	// scheduled onto a different node while the agent on the original node was
+	// down, which would cause the CEP to be deleted once the original agent came
+	// back up. (This holds for StatefulSets in particular that come with stable
+	// pod identifiers and thus do not guard against such accidental deletes
+	// through unique pod names.) Storing the CEP UID at CEP create/fetch time
+	// and using it as a precondition for deletion ensures that agents may only
+	// delete CEPs they own.
+	// It is possible for the CEP UID to not be populated when an agent tries to
+	// clean up a CEP. In that case, skip deletion and rely on cilium operator
+	// garbage collection to clean up eventually.
```

Comment:
This makes sense to me. I took a look at our CEP GC code and it's not quite as "observable" as I'd like if we now more explicitly rely on the GC to get the cluster into a good state. Specifically, we only have a debug level log, which I'd argue should stay that way but we need to augment it.

Could you add a new commit which exposes a metric for how many times we GC a CEP? The code should be around here: https://github.com/cilium/cilium/blob/d783aca9aa277397d60361de0af96da5a4bcf4f0/operator/k8s_cep_gc.go#L149

Reply:
Took a stab at this. Let me know what you think.